File: Rooms.Core/Streams/RoomStream.cs
using System;
using System.IO;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using KolibSoft.Rooms.Core.Protocol;

namespace KolibSoft.Rooms.Core.Streams
{

    /// <summary>
    /// Generic base implementation of a Room stream.
    /// </summary>
    public abstract class RoomStream : IRoomStream
    {

        /// <summary>
        /// Stream options.
        /// </summary>
        public RoomStreamOptions Options { get; private set; } = new RoomStreamOptions();

        /// <summary>
        /// Checks if the stream is ready for read or write messages.
        /// </summary>
        public abstract bool IsAlive { get; }

        /// <summary>
        /// Checks if the instance was disposed.
        /// </summary>
        protected bool IsDisposed => _disposed;

        /// <summary>
        /// Read a chunk of data from the underline implementation.
        /// </summary>
        /// <param name="buffer">Chunk to store the read data.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Amount of bytes read.</returns>
        protected abstract ValueTask<int> ReadAsync(Memory<byte> buffer, CancellationToken token = default);

        /// <summary>
        /// Get the next available chunk of data for read operations.
        /// </summary>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Available chunk or `default` otherwise.</returns>
        private async ValueTask<ReadOnlyMemory<byte>> GetChunkAsync(CancellationToken token = default)
        {
            if (_position == _length)
            {
                _position = 0;
                _length = await ReadAsync(_readBuffer, token);
                if (_length < 1)
                    return default;
            }
            var slice = _readBuffer.AsMemory().Slice(_position, _length - _position);
            return slice;
        }

        /// <summary>
        /// Read a verb.
        /// </summary>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Room verb.</returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask<RoomVerb> ReadVerbAsync(CancellationToken token)
        {
            _data.SetLength(0);
            var done = false;
            while (true)
            {
                var chunk = await GetChunkAsync(token);
                if (_length < 1) throw new IOException("Room verb broken");
                var length = RoomDataUtils.ScanWord(chunk.Span);
                if (length < chunk.Length)
                    length += (done = RoomDataUtils.IsBlank(chunk.Span[length])) ? 1 : 0;
                _position += length;
                if (_data.Length + length > Options.MaxVerbLength) throw new IOException("Room verb too large");
                if (_position < _length || done)
                {
                    if (_data.Length > 0)
                    {
                        await _data.WriteAsync(chunk.Slice(0, length - 1));
                        var verb = new RoomVerb(_data.ToArray());
                        return verb;
                    }
                    if (length > 0)
                    {
                        var verb = new RoomVerb(chunk.Slice(0, length - 1).ToArray());
                        return verb;
                    }
                    return default;
                }
                await _data.WriteAsync(chunk.Slice(0, length));
            }
        }

        /// <summary>
        /// Read a channel.
        /// </summary>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Room channel.</returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask<RoomChannel> ReadChannelAsync(CancellationToken token)
        {
            _data.SetLength(0);
            var done = false;
            while (true)
            {
                var chunk = await GetChunkAsync(token);
                if (_length < 1) throw new IOException("Room channel broken");
                var length = RoomDataUtils.IsSign(chunk.Span[0]) ? 1 : 0;
                if (length < chunk.Length)
                    length += RoomDataUtils.ScanHexadecimal(chunk.Slice(length).Span);
                if (length < chunk.Length)
                    length += (done = RoomDataUtils.IsBlank(chunk.Span[length])) ? 1 : 0;
                _position += length;
                if (_data.Length + length > Options.MaxChannelLength) throw new IOException("Room channel too large");
                if (_position < _length || done)
                {
                    if (_data.Length > 0)
                    {
                        await _data.WriteAsync(chunk.Slice(0, length - 1));
                        var channel = new RoomChannel(_data.ToArray());
                        return channel;
                    }
                    if (length > 0)
                    {
                        var channel = new RoomChannel(chunk.Slice(0, length - 1).ToArray());
                        return channel;
                    }
                    return default;
                }
                await _data.WriteAsync(chunk.Slice(0, length));
            }
        }

        /// <summary>
        /// Read a count.
        /// </summary>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Room count.</returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask<RoomCount> ReadCountAsync(CancellationToken token)
        {
            _data.SetLength(0);
            var done = false;
            while (true)
            {
                var chunk = await GetChunkAsync(token);
                if (_length < 1) throw new IOException("Room count broken");
                var length = RoomDataUtils.ScanDigit(chunk.Span);
                if (length < chunk.Length)
                    length += (done = RoomDataUtils.IsBlank(chunk.Span[length])) ? 1 : 0;
                _position += length;
                if (_data.Length + length > Options.MaxCountLength) throw new IOException("Room count too large");
                if (_position < _length || done)
                {
                    if (_data.Length > 0)
                    {
                        await _data.WriteAsync(chunk.Slice(0, length - 1));
                        var count = new RoomCount(_data.ToArray());
                        return count;
                    }
                    if (length > 0)
                    {
                        var count = new RoomCount(chunk.Slice(0, length - 1).ToArray());
                        return count;
                    }
                    return default;
                }
                await _data.WriteAsync(chunk.Slice(0, length));
            }
        }

        /// <summary>
        /// Read a content.
        /// </summary>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Room content.</returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask<Stream> ReadContentAsync(CancellationToken token)
        {
            var count = await ReadCountAsync(token);
            var _count = (long)count;
            if (_count == 0) return Stream.Null;
            if (_count > Options.MaxContentLength) throw new IOException("Room content too large");
            var content = Options.CreateContentStream(_count);
            var index = 0L;
            while (index < _count)
            {
                var chunk = await GetChunkAsync(token);
                if (_length < 1) throw new IOException("Room content broken");
                var length = (int)Math.Min(chunk.Length, _count - index);
                await content.WriteAsync(chunk.Slice(0, length), token);
                index += length;
                _position += length;
            }
            content.Seek(0, SeekOrigin.Begin);
            return content;
        }

        /// <summary>
        /// Read a message.
        /// </summary>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Room message.</returns>
        /// <exception cref="ObjectDisposedException"></exception>
        public async ValueTask<RoomMessage> ReadMessageAsync(CancellationToken token = default)
        {
            if (_disposed) throw new ObjectDisposedException(nameof(RoomStream));
            var verb = await ReadVerbAsync(token);
            var channel = await ReadChannelAsync(token);
            var content = await ReadContentAsync(token);
            var message = new RoomMessage
            {
                Verb = verb.ToString(),
                Channel = (int)channel,
                Content = content
            };
            return message;
        }

        /// <summary>
        /// Write a chunk of data into the underline implementation.
        /// </summary>
        /// <param name="buffer">Chunk data to write.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns>Amount of bytes written.</returns>
        protected abstract ValueTask<int> WriteAsync(ReadOnlyMemory<byte> buffer, CancellationToken token);

        /// <summary>
        /// Write a verb.
        /// </summary>
        /// <param name="verb">Room verb.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns></returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask WriteVerbAsync(RoomVerb verb, CancellationToken token)
        {
            if (verb.Length > Options.MaxVerbLength) throw new IOException("Room verb too large");
            var index = 0;
            while (index < verb.Length)
            {
                var length = await WriteAsync(verb.Data.Slice(index), token);
                if (length < 1) throw new IOException("Room verb broken");
                index += length;
            }
            await WriteAsync(Blank, token);
        }

        /// <summary>
        /// Write a channel.
        /// </summary>
        /// <param name="channel">Room channel.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns></returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask WriteChannelAsync(RoomChannel channel, CancellationToken token)
        {
            if (channel.Length > Options.MaxChannelLength) throw new IOException("Room channel too large");
            var index = 0;
            while (index < channel.Length)
            {
                var length = await WriteAsync(channel.Data.Slice(index), token);
                if (length < 1) throw new IOException("Room channel broken");
                index += length;
            }
            await WriteAsync(Blank, token);
        }

        /// <summary>
        /// Write a count.
        /// </summary>
        /// <param name="count">Room count.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns></returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask WriteCountAsync(RoomCount count, CancellationToken token)
        {
            if (count.Length > Options.MaxCountLength) throw new IOException("Room count too large");
            var index = 0;
            while (index < count.Length)
            {
                var length = await WriteAsync(count.Data.Slice(index), token);
                if (length < 1) throw new IOException("Room count broken");
                index += length;
            }
            await WriteAsync(Blank, token);
        }

        /// <summary>
        /// Write a content.
        /// </summary>
        /// <param name="content">Room content.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns></returns>
        /// <exception cref="IOException"></exception>
        private async ValueTask WriteContentAsync(Stream content, CancellationToken token)
        {
            if (content.Length > Options.MaxContentLength) throw new IOException("Room content too large");
            var count = (RoomCount)content.Length;
            await WriteCountAsync(count, token);
            content.Seek(0, SeekOrigin.Begin);
            var index = 0L;
            while (index < content.Length)
            {
                var _count = await content.ReadAsync(_writeBuffer, token);
                var slice = _writeBuffer.Slice(0, _count);
                var _index = 0;
                while (_index < slice.Count)
                {
                    var length = await WriteAsync(slice.Slice(_index), token);
                    if (length < 1) throw new IOException("Room content broken");
                    _index += length;
                }
                index += slice.Count;
            }
        }

        /// <summary>
        /// Write a message.
        /// </summary>
        /// <param name="message">Room message.</param>
        /// <param name="token">Cancellation token.</param>
        /// <returns></returns>
        /// <exception cref="ObjectDisposedException"></exception>
        public async ValueTask WriteMessageAsync(RoomMessage message, CancellationToken token = default)
        {
            if (_disposed) throw new ObjectDisposedException(nameof(RoomStream));
            var verb = RoomVerb.Parse(message.Verb);
            var channel = (RoomChannel)message.Channel;
            var content = message.Content;
            await WriteVerbAsync(verb, token);
            await WriteChannelAsync(channel, token);
            await WriteContentAsync(content, token);
        }

        /// <summary>
        /// Dipose implementation.
        /// </summary>
        /// <param name="disposing"></param>
        /// <returns></returns>
        protected virtual async ValueTask OnDisposeAsync(bool disposing)
        {
            if (!_disposed)
            {
                if (disposing)
                    await _data.DisposeAsync();
                _readBuffer = default;
                _writeBuffer = default;
                _disposed = true;
            }
        }

        public void Dispose()
        {
            _ = OnDisposeAsync(disposing: true);
            GC.SuppressFinalize(this);
        }

        public async ValueTask DisposeAsync()
        {
            await OnDisposeAsync(disposing: true);
            GC.SuppressFinalize(this);
        }

        /// <summary>
        /// Initialize the Room stream.
        /// </summary>
        /// <param name="readBuffer">Read buffer.</param>
        /// <param name="writeBuffer">Write buffer.</param>
        /// <param name="options">Strea options.</param>
        protected RoomStream(ArraySegment<byte> readBuffer, ArraySegment<byte> writeBuffer, RoomStreamOptions? options = null)
        {
            _readBuffer = readBuffer;
            _writeBuffer = writeBuffer;
            Options = options ?? new RoomStreamOptions();
        }

        /// <summary>
        /// Initialize the Room stream.
        /// </summary>
        /// <param name="options">Stream options.</param>
        protected RoomStream(RoomStreamOptions? options = null)
        {
            Options = options ?? new RoomStreamOptions();
            _readBuffer = new byte[Options.ReadBuffering];
            _writeBuffer = new byte[Options.WriteBuffering];
        }

        /// <summary>
        /// Internal message buffer.
        /// </summary>
        private MemoryStream _data = new MemoryStream();

        /// <summary>
        /// Internal read buffer.
        /// </summary>
        private ArraySegment<byte> _readBuffer = default;

        /// <summary>
        /// Internal write buffer.
        /// </summary>
        private ArraySegment<byte> _writeBuffer = default;

        /// <summary>
        /// Internal reading position.
        /// </summary>
        private int _position = 0;

        /// <summary>
        /// Internal reading length.
        /// </summary>
        private int _length = 0;

        /// <summary>
        /// Internal dispose flag.
        /// </summary>
        private bool _disposed = false;

        /// <summary>
        /// Internal blank data.
        /// </summary>
        private static readonly byte[] Blank = Encoding.UTF8.GetBytes(" ");

    }

}

File: Rooms.Web/wwwroot/src/streams/room_stream.js
import { RoomChannel } from "../protocol/room_channel.js";
import { RoomCount } from "../protocol/room_count.js";
import { RoomDataUtils, encoder } from "../protocol/room_data_utils.js";
import { RoomMessage } from "../protocol/room_message.js";
import { RoomVerb } from "../protocol/room_verb.js";
import { MemoryStream } from "./memory_stream.js";
import { RoomStreamOptions } from "./room_stream_options.js";

const BLANK = encoder.encode(" ");

class RoomStream {

    #options;
    #data;
    #readBuffer;
    #writeBuffer;
    #position;
    #length;
    #disposed;

    get options() { return this.#options; }
    get isAlive() { throw Error("Abstract member"); }
    get isDisposed() { return this.#disposed; }

    async readAsync(chunk) { throw Error("Abstract member"); }

    async #getChunkAsync() {
        if (this.#position == this.#length) {
            this.#position = 0;
            this.#length = await this.readAsync(this.#readBuffer);
            if (this.#length < 1)
                return new Uint8Array(0);
        }
        let slice = RoomDataUtils.slice(this.#readBuffer, this.#position, this.#length - this.#position);
        return slice;
    }

    async #readVerbAsync() {
        this.#data.position = 0;
        this.#data.length = 0;
        let done = false;
        while (true) {
            let chunk = await this.#getChunkAsync();
            if (this.#length < 1) throw new Error("Room verb broken");
            let length = RoomDataUtils.scanWord(chunk);
            if (length < chunk.length)
                length += (done = RoomDataUtils.isBlank(chunk[length])) ? 1 : 0;
            this.#position += length;
            if (this.#data.length + length > this.#options.maxVerbLength) throw new Error("Room verb too large");
            if (this.#position < length || done) {
                if (this.#data.length > 0) {
                    await this.#data.writeAsync(RoomDataUtils.slice(chunk, 0, length - 1));
                    let verb = new RoomVerb(this.#data.toArray());
                    return verb;
                }
                if (length > 0) {
                    let verb = new RoomVerb(chunk.slice(0, length - 1));
                    return verb;
                }
                return new RoomVerb();
            }
            await this.#data.writeAsync(chunk);
        }
    }

    async #readChannelAsync() {
        this.#data.position = 0;
        this.#data.length = 0;
        let done = false;
        while (true) {
            let chunk = await this.#getChunkAsync();
            if (this.#length < 1) throw new Error("Room channel broken");
            let length = RoomDataUtils.isSign(chunk[0]) ? 1 : 0;
            if (length < chunk.length)
                length += RoomDataUtils.scanHexadecimal(RoomDataUtils.slice(chunk, length));
            if (length < chunk.length)
                length += (done = RoomDataUtils.isBlank(chunk[length])) ? 1 : 0;
            this.#position += length;
            if (this.#data.length + length > this.#options.maxChannelLength) throw new Error("Room channel too large");
            if (this.#position < length || done) {
                if (this.#data.length > 0) {
                    await this.#data.writeAsync(RoomDataUtils.slice(chunk, 0, length - 1));
                    let channel = new RoomChannel(this.#data.toArray());
                    return channel;
                }
                if (length > 0) {
                    let channel = new RoomChannel(chunk.slice(0, length - 1));
                    return channel;
                }
                return new RoomChannel();
            }
            await this.#data.writeAsync(chunk);
        }
    }

    async #readCountAsync() {
        this.#data.position = 0;
        this.#data.length = 0;
        let done = false;
        while (true) {
            let chunk = await this.#getChunkAsync();
            if (this.#length < 1) throw new Error("Room count broken");
            let length = RoomDataUtils.scanDigit(chunk);
            if (length < chunk.length)
                length += (done = RoomDataUtils.isBlank(chunk[length])) ? 1 : 0;
            this.#position += length;
            if (this.#data.length + length > this.#options.maxCountLength) throw new Error("Room count too large");
            if (this.#position < length || done) {
                if (this.#data.length > 0) {
                    await this.#data.writeAsync(RoomDataUtils.slice(chunk, 0, length - 1));
                    let count = new RoomCount(this.#data.toArray());
                    return count;
                }
                if (length > 0) {
                    let count = new RoomCount(chunk.slice(0, length - 1));
                    return count;
                }
                return new RoomCount();
            }
            await this.#data.writeAsync(chunk);
        }
    }

    async #readContentAsync() {
        let count = await this.#readCountAsync();
        let _count = parseInt(count);
        if (_count == 0) return {};
        if (_count > this.#options.maxContentLength) throw new Error("Room content too large");
        let content = new MemoryStream();
        let index = 0;
        while (index < _count) {
            let chunk = await this.#getChunkAsync();
            if (this.#length < 1) throw new Error("Room content broken");
            let length = Math.min(chunk.length, _count - index);
            await content.writeAsync(RoomDataUtils.slice(chunk, 0, length));
            index += length;
            this.#position += length;
        }
        return content.toArray();
    }

    async readMessageAsync() {
        if (this.#disposed) throw new Error("RoomStream was disposed");
        let verb = await this.#readVerbAsync();
        let channel = await this.#readChannelAsync();
        let content = await this.#readContentAsync();
        let message = new RoomMessage({
            verb: verb.toString(),
            channel: parseInt(channel, 16),
            content
        });
        return message;
    }

    async writeAsync(chunk) { throw Error("Abstract member"); }

    async #writeVerbAsync(verb) {
        if (!(verb instanceof RoomVerb)) throw new Error("Invalid argument");
        if (verb.length > this.#options.maxVerbLength) throw new Error("Room verb too large");
        let index = 0;
        while (index < verb.length) {
            let length = await this.writeAsync(RoomDataUtils.slice(verb.data, index));
            if (length < 1) throw new Error("Room verb broken");
            index += length;
        }
        await this.writeAsync(BLANK);
    }

    async #writeChannelAsync(channel) {
        if (!(channel instanceof RoomChannel)) throw new Error("Invalid argument");
        if (channel.length > this.#options.maxChannelLength) throw new Error("Room channel too large");
        let index = 0;
        while (index < channel.length) {
            let length = await this.writeAsync(RoomDataUtils.slice(channel.data, index));
            if (length < 1) throw new Error("Room channel broken");
            index += length;
        }
        await this.writeAsync(BLANK);
    }

    async #writeCountAsync(count) {
        if (!(count instanceof RoomCount)) throw new Error("Invalid argument");
        if (count.length > this.#options.maxChannelLength) throw new Error("Room count too large");
        let index = 0;
        while (index < count.length) {
            let length = await this.writeAsync(RoomDataUtils.slice(count.data, index));
            if (length < 1) throw new Error("Room count broken");
            index += length;
        }
        await this.writeAsync(BLANK);
    }

    async #writeContentAsync(content) {
        if (!(content instanceof Uint8Array)) throw new Error("Invalid argument");
        if (content.length > this.#options.maxContentLength) throw new Error("Room content too large");
        let count = new RoomCount(content.length);
        await this.#writeCountAsync(count);
        let index = 0;
        while (index < content.length) {
            this.#writeBuffer.set(content);
            let _count = content.length;
            let slice = RoomDataUtils.slice(this.#writeBuffer, 0, _count);
            let _index = 0;
            while (_index < slice.length) {
                let length = await this.writeAsync(RoomDataUtils.slice(slice, _index));
                if (length < 1) throw new Error("Room content broken");
                _index += length;
            }
            index += slice.length;
        }
    }

    async writeMessageAsync(message) {
        if (!(message instanceof RoomMessage)) throw new Error("Invalid argument");
        if (this.#disposed) throw new Error("RoomStream was disposed");
        let verb = RoomVerb.parse(message.verb);
        let channel = new RoomChannel(message.channel);
        let content = message.content;
        await this.#writeVerbAsync(verb);
        await this.#writeChannelAsync(channel);
        await this.#writeContentAsync(content);
    }

    async onDisposeAsync(disposing) {
        if (!this.#disposed) {
            if (disposing)
                this.#data.length = 0;
            this.#readBuffer = null;
            this.#writeBuffer = null;
            this.#disposed = true;
        }
    }

    dispose() { let _ = this.onDisposeAsync(true); }
    async disposeAsync() { await this.onDisposeAsync(true); }

    constructor(args = { readBuffer: null, writeBuffer: null, options: null }) {
        if (args?.readBuffer && !(args.readBuffer instanceof Uint8Array) || args?.writeBuffer && !(args.writeBuffer instanceof Uint8Array))
            throw new Error("Invalid arguments");
        this.#options = new RoomStreamOptions(args?.options);
        this.#readBuffer = args?.readBuffer ?? new Uint8Array(this.#options.readBuffering);
        this.#writeBuffer = args?.writeBuffer ?? new Uint8Array(this.#options.writeBuffering);
        this.#data = new MemoryStream();
        this.#position = 0;
        this.#length = 0;
        this.#disposed = false;
    }

}

export {
    RoomStream
}